specs/agents/tracing-spans-compress.md

```diff
@@ -0,0 +1,211 @@
+APM Server tracks span destination metrics.
+To avoid compressed spans to skew latency metrics and cause throughput metrics to be under-counted,
+APM Server will take `composite.count` into account when tracking span destination metrics.
+
+### Compression algorithm
+
+#### Eligibility for compression
+
+A span is eligible for compression if all the following conditions are met
+- It's an [exit span](tracing-spans.md#exit-spans)
+- The trace context of this span has not been propagated to a downstream service
+
+The latter condition is important so that we don't remove (compress) a span that may be the parent of a downstream service.
+This would orphan the sub-graph started by the downstream service and cause it to not appear in the waterfall view.
+
+```java
+boolean isCompressionEligible() {
+    return exit && !context.hasPropagated
+}
+```
+
+#### Span buffering
+
+Non-compression-eligible spans may be reported immediately after they have ended.
+When a compression-eligible span ends, it does not immediately get reported.
+Instead, the span is buffered within its parent.
+A span/transaction can buffer at most one child span.
+
+Span buffering allows to "look back" one span when determining whether a given span should be compressed.
+
+A buffered span gets reported when
+1. its parent ends
+2. a non-compressible sibling ends
+
+```java
+void onSpanEnd() {
+    if (isCompressionEligible()) {
+        if (parent.hasBufferedSpan()) {
+            parent.tryCompress(this)
+        } else {
+            parent.buffered = this
+        }
+    } else { 
+        report(buffered)
+        report(this)
+    }
+}
+```
+
+#### Compression
+
+On span end, we compare each [compression-eligible](tracing-spans-compress.md#eligibility-for-compression) span to it's previous sibling.
+
+If the spans are of the same kind but have different span names and the compressions-eligible span's `duration` <= `same_kind_compression_max_duration`,
+we compress them using the [Consecutive-Same-Kind compression strategy](tracing-spans-compress.md#consecutive-same-kind-compression-strategy).
+
+If the spans are of the same kind, and have the same name,
+we apply the [Consecutive-Exact-Match compression strategy](tracing-spans-compress.md#consecutive-exact-match-compression-strategy).
+
+```java
+void tryCompress(Span child) {
+    if (buffered.isSameKind(child)) {
+        if (buffered.name == child.name) {
+            buffered.compress(child, exactMatch: true)
+            return
+        } else if ( (buffered.duration <= same_kind_compression_max_duration || buffered.composite.count > 1)
+                   && child.duration <= same_kind_compression_max_duration) {
+            buffered.name = "Calls to $buffered.destination.service.resource"
+            buffered.compress(child, exactMatch: false)
+            return
+        }
+    }
+    report(buffered)
+    buffered = child
+}
```

Comment:
```suggestion
void tryCompress(Span child) {
    if (tryToAddToBuffered(child)) {
        return
    }

    reportAndClear(buffered)
    buffered.add(child)
}

bool tryToAddToBuffered(Span child) {
    if (buffered.isEmpty()) {
        buffered.add(child)
        buffered.firstChild = child
        return true
    }

    if (!buffered.isSameKind(child)) {
        return false
    }

    if (buffered.count == 1) {
        if (buffered.firstChild.name == child.name) {
            buffered.exactMatch = true
            buffered.add(child)
            buffered.name = child.name
            return true
        }

        if (buffered.firstChild.duration <= same_kind_compression_max_duration && child.duration <= same_kind_compression_max_duration) {
            buffered.exactMatch = false
            buffered.add(child)
            buffered.name = "Calls to $buffered.destination.service.resource"
            return true
        }

        return false
    }

    if (buffered.exactMatch) {
        if (buffered.name == child.name) {
            buffered.add(child)
            return true
        }

        return false
    }

    if (child.duration <= same_kind_compression_max_duration) {
        buffered.add(child)
        return true
    }

    return false
}
```

Reply:
I didn't get a chance to fully process this before my PTO. Just some initial reactions `buffered` was meant to be a regular span. The first time another span gets `compress`ed, it turns into a composite span. See also https://github.com/felixbarny/apm/blob/compressed-spans/specs/agents/tracing-spans-compress.md#turning-compressed-spans-into-a-composite-span.

I think the `buffered.add(child)` from your proposal is similar to what I had in mind with the compress method. But add suggests that the span is added into a list which is not what I had in mind.

Reply:
> I think the `buffered.add(child)` from your proposal is similar to what I had in mind with the compress method. But add suggests that the span is added into a list which is not what I had in mind.

You're right - usage `buffered.add(child)` might be confusing. I changed it `buffered.compress(child)` and refactored a little bit.